great_expectations/datasource/fluent/sources.py

```diff
@@ -118,44 +123,101 @@ def register_types_and_ds_factory(
 
             cls._register_assets(ds_type, asset_type_lookup=asset_type_lookup)
 
-            cls._register_datasource_and_factory_method(
+            cls._register_datasource(
                 ds_type,
-                factory_fn=factory_fn,
                 ds_type_name=ds_type_name,
                 datasource_type_lookup=ds_type_lookup,
             )
 
     @classmethod
-    def _register_datasource_and_factory_method(
+    def _register_datasource(
         cls,
         ds_type: Type[Datasource],
-        factory_fn: SourceFactoryFn,
         ds_type_name: str,
         datasource_type_lookup: TypeLookup,
     ) -> str:
         """
         Register the `Datasource` class and add a factory method for the class on `sources`.
         The method name is pulled from the `Datasource.type` attribute.
         """
-        method_name = f"add_{ds_type_name}"
-        logger.debug(
-            f"2a. Registering {ds_type.__name__} as {ds_type_name} with {method_name}() factory"
-        )
-
-        pre_existing = cls.__source_factories.get(method_name)
-        if pre_existing:
+        if ds_type in datasource_type_lookup:
             raise TypeRegistrationError(
-                f"'{ds_type_name}' - `sources.{method_name}()` factory already exists",
+                f"'{ds_type_name}' is already a registered typed and there can only be 1 type "
+                "for a given name."
             )
-
         datasource_type_lookup[ds_type] = ds_type_name
         logger.debug(f"'{ds_type_name}' added to `type_lookup`")
-        factory_fn.__name__ = method_name
-        public_api(factory_fn)
 
-        cls.__source_factories[method_name] = factory_fn
+        cls._register_add_datasource(ds_type, ds_type_name)
+        cls._register_update_datasource(ds_type, ds_type_name)
+        cls._register_add_or_update_datasource(ds_type, ds_type_name)
+        cls._register_delete_datasource(ds_type, ds_type_name)
+
         return ds_type_name
 
+    @classmethod
+    def _register_add_datasource(cls, ds_type: Type[Datasource], ds_type_name: str):
+        method_name = f"add_{ds_type_name}"
+
+        def crud_method_info() -> tuple[CrudMethodType, Type[Datasource]]:
+            return CrudMethodType.ADD, ds_type
+
+        cls._register_crud_method(method_name, cls.__doc__, crud_method_info)
+
+    @classmethod
+    def _register_update_datasource(cls, ds_type: Type[Datasource], ds_type_name: str):
+        method_name = f"update_{ds_type_name}"
+
+        def crud_method_info() -> tuple[CrudMethodType, Type[Datasource]]:
```

Comment:
Non-blocking

Might just need to read more but why is this a closure instead of just passing the tuple or using an inline lambda?

Reply:
Good idea! This could be a tuple instead of a closure or a lambda. I'll update this.

Reply:
Oh, I made it a callable because I needed something to wrap the `@public_api` decorator around and this is a proxy for the what eventually gets called by the crud method.